kpi-tracker/src/main.rs

```diff
@@ -702,30 +765,131 @@ async fn use_node(db: &mut DB, from_height: AbsoluteBlockHeight) -> anyhow::Resu
         .context("Could not connect to node.")?;
 
     let mut blocks_stream = node
-        .get_finalized_blocks_from(from_height)
+        .get_finalized_blocks_from(latest_height)
         .await
         .context("Error querying blocks")?;
 
-    let start_height = if from_height.height == 0 {
+    if latest_height.height == 0 {
         if let Some(genesis_block) = blocks_stream.next().await {
-            process_genesis_block(&mut node, genesis_block.block_hash, db).await?;
+            let genesis_block_data =
+                process_genesis_block(&mut node, genesis_block.block_hash).await?;
+            block_sender
+                .send(BlockData::Genesis(genesis_block_data))
+                .await?;
             log::info!("Processed genesis block: {}", genesis_block.block_hash);
         }
-        from_height.height + 1
-    } else {
-        from_height.height
     };
 
-    for height in start_height..blocks_to_process {
+    for height in latest_height.height.. {
         if let Some(block) = blocks_stream.next().await {
-            process_block(&mut node, block.block_hash, db).await?;
+            let block_data = process_block(&mut node, block.block_hash).await?;
+            block_sender.send(BlockData::Normal(block_data)).await?;
             log::info!("Processed block ({}): {}", height, block.block_hash);
         }
     }
```

Comment:
The outer for loop does not seem useful and is kind of wrong.

You should get the block height from the block stream directly. That way you will easily make sure it is correct. As opposed to how where you have to match two separate streams.